### PR TITLE
Add support for generic persistence interface -- alternative 1

### LINKED ISSUE #23: Consider supporting a generic persistence interface
Use cases:
1. Custom wrappers over SharedPreferences -- some projects abstract away from persistence, making this package a bad choice for them, since we only support stock SharedPreferences.
2. Secure storage -- while iOS should be good to go, [`sharedpreferences` on Android is not the most secure way to store data](https://stackoverflow.com/a/9244620/12461510). Adding a generic interface should give the user freedom of choice when it comes to choosing the layer which does the actual key/value persistence.

In general the idea is to abstract away from details on how we do persistence, and turn this package into a generic reactive extension for a keyed storage.

### PR DESCRIPTION
**Pull request by solid-yuriiprykhodko**
Resolves #23 
Alternative to #26 

Abstracts away only the key-value store.
The persistence `key` and also the `hydrate` and `persist` callbacks are kept in the `HydratedSubject` class, avoiding breaking changes.

### CODE REVIEW

File: lib/src/hydrated.dart [hunk 1]
```diff
@@ -58,21 +47,23 @@ typedef HydrateCallback<T> = T Function(String);
...
 /// ```
 class HydratedSubject<T> extends Subject<T> implements ValueStream<T> {
-  static final _areTypesEqual = TypeUtils.areTypesEqual;
-  final BehaviorSubject<T> _subject;
   final String _key;
   final HydrateCallback<T>? _hydrate;
   final PersistCallback<T>? _persist;
+  final BehaviorSubject<T> _subject;
   final VoidCallback? _onHydrate;
   final T? _seedValue;
```
Comment (solid-software) [inline lib/src/hydrated.dart:56]:
```suggestion
```

File: lib/src/hydrated.dart [hunk 2]
```diff
@@ -122,13 +100,10 @@ class HydratedSubject<T> extends Subject<T> implements ValueStream<T> {
...
       onHydrate,
       subject,
+      persistence,
     );
   }
 
-  /// A unique key that references a storage container
-  /// for a value persisted on the device.
-  String get key => _key;
-
```
Comment (solid-software) [inline lib/src/hydrated.dart:131]: Could this be considered as breaking change? Do we need to remove it?

File: lib/src/hydrated.dart [hunk 3]
```diff
@@ -163,72 +138,47 @@ class HydratedSubject<T> extends Subject<T> implements ValueStream<T> {
...
   /// Hydrates the HydratedSubject with a value stored on the user's device.
   ///
   /// Must be called to retrieve values stored on the device.
   Future<void> _hydrateSubject() async {
-    final prefs = await SharedPreferences.getInstance();
-
-    T? val;
+    try {
+      T? val;
+      if (_doEncodePersistedValue) {
```
Comment (solid-software) [inline lib/src/hydrated.dart:149]:
```suggestion
      if (_hydrate != null) {
```
maybe that would be clearer?

File: lib/src/hydrated.dart [hunk 4]
```diff
@@ -163,72 +138,47 @@ class HydratedSubject<T> extends Subject<T> implements ValueStream<T> {
...
+        final encodedValue = await _persistence.get<String>(_key);
```
Comment (solid-software) [inline lib/src/hydrated.dart:150]:
```suggestion
        final storedValue = await _persistence.get<String>(_key);
```
?

File: lib/src/persistence/persistence_error.dart [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class PersistenceError extends Error {
```
Comment (solid-software) [inline lib/src/persistence/persistence_error.dart:1]: Could you please add documentation for this class?

File: lib/src/persistence/persistence_error.dart [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+  final String? message;
```
Comment (solid-software) [inline lib/src/persistence/persistence_error.dart:2]:
```suggestion
  final String message;
```
What do you think?

**Review by ghost — CHANGES_REQUESTED** [review]
Nicely done @solid-yuriiprykhodko! I've added a couple of comments - please review and let me know what you think. Great addition to the hydrated plugin.

File: lib/src/hydrated.dart (hunk 1 above)
Comment (solid-yuriiprykhodko) [reply lib/src/hydrated.dart:56]: Not sure this is possible -- we are delegating several methods to `_subject` in order to conform to the `Subject<T>` interface.

Unfortunately, it is not possible to extend `BehaviorSubject` directly, so we are forced to create tramp methods, such as this:

```dart
  @override
  bool get hasValue => _subject.hasValue;
```

File: lib/src/hydrated.dart (hunk 4 above)
Comment (solid-yuriiprykhodko) [reply lib/src/hydrated.dart:150]: Went with `persistedValue`.

File: lib/src/persistence/persistence_error.dart (hunk 6 above)
Comment (solid-yuriiprykhodko) [reply lib/src/persistence/persistence_error.dart:2]: Agree! Suggestion applied manually.

File: lib/src/persistence/persistence_error.dart (hunk 5 above)
Comment (solid-yuriiprykhodko) [reply lib/src/persistence/persistence_error.dart:1]: Done.

File: lib/src/hydrated.dart (hunk 3 above)
Comment (solid-yuriiprykhodko) [reply lib/src/hydrated.dart:149]: Agreed. Suggestion applied manually.

File: lib/src/hydrated.dart (hunk 2 above)
Comment (solid-yuriiprykhodko) [reply lib/src/hydrated.dart:131]: Restored.

File: lib/hydrated.dart [hunk 7]
```diff
@@ -1,3 +1,6 @@
 library hydrated;
 
 export 'src/hydrated.dart';
+export 'src/persistence/key_value_store.dart';
```
Comment (solid-yuriiprykhodko) [inline lib/hydrated.dart:4]: test

**Review by ghost — APPROVED** [review]
LGTM - thanks @solid-yuriiprykhodko !

**Pull request merged**